Dragons drop devoured bodies, no stomach attacks and passive regeneration

Arendian opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->

resolves #21449
resolves #23540
resolves #20935
resolves #20405
resolves #16726
resolves #10816

Adds a Devoured component to any devoured entities that passively deals caustic damage while an entity has it.
The damage is based on the entities MobState, with it being the highest while the entity is alive and the lowest while the entity is dead.
Damage dealt in this way is currently capped at 100 damage more than when an entity is considered dead.
Entities get the Pacified component while devoured.

Devoured bodies won't be deleted when the dragon is gibbed, butchered or when the dragon despawns from not making enough rifts.
After being devoured entities will stop bleeding, this is to prevent the dragon leaving a trail of blood puddles behind them.

Devoured entities won't rot, not too sure about this one because it does remove a lot of the "danger feeling" a dragon has if everyone can just have a happy ending the moment the dragon is killed.

Dragons now have a low amount of passive healing, and  in exchange no ichor gets injected anymore when they devour a humanoid entity.
Devouring a humanoid entity permanently increases their passive healing.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
Entities in a dragon's stomach should take damage from the stomach acids it (probably?) has.
The damage is lowered on entities in a critical or dead state so that a timely retrieval of the body will need less healing or may even save them before they die.

A dragon can't defend itself against enemies that are inside their stomach, making the devoured entity pacified prevents this.

If the crew manages to kill the dragon they should be able to retrieve the devoured bodies for reviving/burial/borging.

Having the passive healing increase for every devoured entity will reward the dragon player if they manage to devour a lot of crew, and can develop them into an even more dangerous threat if the crew doesn't manage to stop the dragon early on.
 Some feedback on the amount of healing would be appreciated.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
When a space dragon devours an entity, the entity will get the Devoured, Pacified and PassiveDamage component if they don't have these yet. These are removed again when the entities are released from the stomach.

The devoured component passively deals damage to an entity that has it, with a multiplier for this damage based on the entities MobState.

Before the dragon gets deleted the stomach will be emptied.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->
Video is outdated, no chemicals are injected anymore, but the current version still achieves the same result.

https://github.com/space-wizards/space-station-14/assets/137322659/fbf1309f-ecef-4a47-95c2-28643c755c9b


- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl: Dygon
- add: Space dragons now have passive health regeneration that increases with every humanoid they devour.
- remove: Space dragons don't get an ichor injection anymore after devouring a humanoid.
- tweak: Bodies devoured by a space dragon don't rot.
- fix: Mobs devoured by a space dragon now won't be able to attack and will continuously take damage.
- fix: Bodies devoured by a space dragon won't be deleted anymore when the dragon is butchered.

metalgearsloth commented on Content.Server/Devour/DevourSystem.cs:
```diff
@@ -32,6 +44,9 @@ private void OnDoAfter(EntityUid uid, DevourerComponent component, DevourDoAfter
             if (component.ShouldStoreDevoured && args.Args.Target is not null)
             {
                 ContainerSystem.Insert(args.Args.Target.Value, component.Stomach);
+
+                //Stops the bleeding of the devoured target so the dragon doesn't leave a trail of random blood puddles behind
+                _bloodstreamSystem.TryModifyBleedAmount(args.Args.Target.Value, -100);
             }
```
This should be fixed on the bloodstream end and not worked around externally.

Arendian replied:
I'm not 100% sure I understood what you meant, does the event I made come close to it?

metalgearsloth commented on Content.Server/Devour/DevourSystem.cs:
```diff
@@ -45,5 +60,19 @@ private void OnDoAfter(EntityUid uid, DevourerComponent component, DevourDoAfter
         _audioSystem.PlayPvs(component.SoundDevour, uid);
     }
+
+    private void EmptyStomach(EntityUid uid, DevourerComponent component)
+    {
+        //Remove entities from the stomach before deletion.
+        foreach (var entity in component.Stomach.ContainedEntities.ToArray())
+        {
+            _containerSystem.Remove(entity, component.Stomach);
+        }
+    }
```
Isn't this just EmptyContainer?

Arendian replied:
Apparently, yes!

metalgearsloth commented on Content.Shared/Devour/Components/DevourerComponent.cs:
```diff
@@ -71,6 +72,18 @@ public sealed partial class DevourerComponent : Component
     [ViewVariables(VVAccess.ReadWrite), DataField("healRate")]
     public float HealRate = 15f;
 
+    /// <summary>
+    /// The chemical ID's injected into the victim upon being devoured.
+    /// </summary>
+    [DataField("stomachChemicals", customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>))]
+    public List<string> StomachChemicals = new();
```
Please update to modern codebase standards i.e. no datafield string and using protoid.

Arendian replied:
I'll do a reread of the new conventions before I will make any new PR's

metalgearsloth commented on Content.Shared/Devour/Components/DevourerComponent.cs:
```diff
@@ -71,6 +72,18 @@ public sealed partial class DevourerComponent : Component
+    [DataField("stomachChemicals", customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>))]
+    public List<string> StomachChemicals = new();
+
+    /// <summary>
+    /// The amount of stomach chemicals  injected into the devoured entity.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("damageRate")]
+    public float DamageRate = 50f;
```
DamageRate does not correspond to what this is actually doing as it has nothing to do with damage.

You're better off just using a solution instead.

Arendian replied:
I tried following the naming already used in the component, the datafield keeping track of the amount of ichor injected was called HealRate.

Kadeo64 commented:
I'm against the whole injecting carpotoxin into people's bodies, I think pax is enough. The crit gasping makes it obvious the dragon has devoured someone and is currently healing, I think pax is all the dragon needs.

Kadeo64 commented:
also devoured entities not rotting is a good change, because it's annoying for people to rot when the dragon devours three people and wild goose chases for 15 minutes.

Arendian commented:
> edit: forgor zombies existed

The carpotoxin is mainly for the zombies yes, also fits the theme of the dragon since it's distant family of the space carps.

Arendian commented:
I'm converting the injection parts to a system that adds pacifieComponent and passiveDamageComponent

Arendian commented:
I added a system that applies pacified and a passive damage effect to any devoured entities. I also replaced the ichor injection with a passive health regeneration that increases with every humanoid devoured.

Arendian commented:
The passive healing change might be out of scope for this PR so I can put it in another one if that makes this one easier to review,

deltanedas commented:
> Devouring a humanoid entity permanently increases their passive healing.

juggernaut real

i like no rotting in stomach since once you open dragon up the rot timer begins and med has to run around like mad to try morgue/bag and heal everyone

deltanedas commented on Content.Server/Body/Systems/BloodstreamSystem.cs:
```diff
@@ -458,7 +458,7 @@ public void ChangeBloodReagent(EntityUid uid, string reagent, BloodstreamCompone
     ///     Stops the bleeding of the devoured target so the dragon doesn't leave a trail of random blood puddles behind
     /// </summary>
-    private void OnDevoured(EntityUid uid, BloodstreamComponent component, OnDevouredEvent args)
+    private void OnDevoured(EntityUid uid, BloodstreamComponent component, ref OnDevouredEvent args)
```
Entity<BloodstreamComponent> ent and use ent.Comp to get the component, do this for all stuff

deltanedas commented on Content.Server/Devour/DevouredSystem.cs:
```diff
@@ -24,42 +24,42 @@ public override void Initialize()
         //the damage cap specified in the component. This can be used to make reviving take more work if
         //the target has been devoured for a long time.
-        if (TryComp<MobThresholdsComponent>(uid, out var mobThresholds))
+        if (TryComp<MobThresholdsComponent>(entity.Owner, out var mobThresholds))
```
Entity automatically casts to EntityUid so just pass entity to trycomp

metalgearsloth commented on Content.Server/Devour/DevouredSystem.cs:
```diff
@@ -0,0 +1,136 @@
+                passiveDamage.Damage.TrimZeros();
+            }
+        }
+
+        //Remove the pacified component if the target originally didn't have it.
+        if (!component.OriginallyPacified)
+            RemComp<PacifiedComponent>(uid);
```
Why is this needed?

Arendian replied:
It's so people that had the pacified component before they got devoured(thieves and people with the pacifist trait) will still have the component. But I'll change it to actionblocker like you suggested.

metalgearsloth commented:
Did a partial review

Arendian commented on Content.Server/Devour/Components/DevouredComponent.cs:
```diff
@@ -0,0 +1,68 @@
+    /// </summary>
+    [DataField]
+    public List<MobState>? OriginalAllowedMobStates;
+
```
I fixed the prediction issue

AJCM-git commented on Content.Server/Dragon/DragonSystem.cs:
```diff
@@ -98,7 +100,9 @@ public override void Update(float frameTime)
             {
                 Roar(uid, comp);
-                QueueDel(uid);
+                if(TryComp<DevourerComponent>(uid, out var devourComponent))
+                    _devour.EmptyStomach(uid, devourComponent);
+                _body.GibBody(uid);
```
wouldnt gibbing already empty the stomach since it would raise the event that gets handled by the devour system?

AJCM-git commented on Content.Server/Devour/DevourSystem.cs:
```diff
@@ -1,49 +1,89 @@
+    /// <summary>
+    /// Empties the stomach when component is removed.
+    /// </summary>
+    private void OnRemoved(EntityUid uid, DevourerComponent component, ComponentRemove args)
```
i could be misremembering but i could swear gibbing raised ComponentRemove since it deletes the entity, is the other sub necessary?

AJCM-git commented on Content.Server/Zombies/ZombieSystem.Transform.cs:
```diff
@@ -116,7 +116,13 @@ public void ZombifyEntity(EntityUid target, MobStateComponent? mobState = null)
             var combat = EnsureComp<CombatModeComponent>(target);
-            RemComp<PacifiedComponent>(target);
+
+            //Only remove the pacified component if the target hasn't been devoured.
+            //This is to prevent entities that turn into a zombie while devoured
+            //from attacking the entity that devoured them from the inside.
+            if(!HasComp<DevouredComponent>(target))
+                RemComp<PacifiedComponent>(target);
```
dont you cancel the attack attempt now?

AJCM-git commented on Content.Shared/Devour/Components/DevouredComponent.cs:
```diff
@@ -0,0 +1,54 @@
+    ///     This is so it can be returned to it's original state when the component is removed.
+    /// </summary>
+    [DataField]
+    public List<MobState>? OriginalAllowedMobStates;
```
I really dislike storing this, is it really necessary?

AJCM-git commented on Content.Server/Devour/DevourSystem.cs:
```diff
@@ -1,49 +1,89 @@
-                ContainerSystem.Insert(args.Args.Target.Value, component.Stomach);
+                if (component.PassiveDevourHealing != null)
+                {
+                    passiveHealing.Damage += component.PassiveDevourHealing;
```
how many bodies do a dragon stomach fit?

Kadeo64 replied:
i'm like 90% sure the dragon doesn't have a coded limit but if you devour the integer limit's worth of bodies it'll probably cause a server crash

AJCM-git replied:
then either there needs to be a cap in the stomach or in the healing factor

AJCM-git commented on Content.Shared/Devour/Components/DevouredComponent.cs:
```diff
@@ -0,0 +1,54 @@
+    /// </summary>
+    public DamageSpecifier? StomachDamage;
+
+    /// <summary>
+    ///     How much extra damage can be done to the entity after it has died.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 DamageCap = 100;
+
+    /// <summary>
+    ///     Stores the current damage modifier.
+    ///     This is to make sure the correct amount of damage is changed.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 CurrentModifier;
+
+    /// <summary>
+    ///     Damage multiplier if the devoured entity is alive.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 AliveMultiplier = 5;
+
+    /// <summary>
+    ///     Damage multiplier if the devoured entity is critical.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 CritMultiplier = 1;
+
+    /// <summary>
+    ///     Damage multiplier if the devoured entity is dead.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 DeadMultiplier = 0.25;
```
Should prob just apply the same amount of damage while you are alive and only stop once dead to make the code and mechanics more straight-forward

AJCM-git commented on Content.Shared/Devour/Components/DevouredComponent.cs:
```diff
@@ -0,0 +1,54 @@
+    /// <summary>
+    ///     Stores the stomach damage so it can be deducted from the devoured entity when the damage is altered.
+    /// </summary>
+    public DamageSpecifier? StomachDamage;
```
datafield?

AJCM-git commented on Content.Shared/Devour/Components/DevourerComponent.cs:
VV.RW is now inferred so you can put only datafield

AJCM-git commented:
@Arendian you coming back to this?

deltanedas commented:
calling dibs on this

deltanedas commented:
this ded
